File: src/entities/Wall.cpp
/*
 * Copyright (C) 2006-2014 Christopho, Solarus - http://www.solarus-games.org
 * 
 * Solarus is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Solarus is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
#include "entities/Wall.h"
#include "entities/Hero.h"
#include "lowlevel/FileTools.h"

namespace solarus {

/**
 * \brief Constructor.
 * \param name name of the wall to create
 * \param layer layer of the Wall to create
 * \param x x position of the Wall to create
 * \param y y position of the Wall to create
 * \param width width of the Wall to create
 * \param height height of the Wall to create
 * \param stops_hero true to make this entity an obstacle for the hero
 * \param stops_enemies true to make this entity an obstacle for the enemies
 * \param stops_npcs true to make this entity an obstacle for the NPCs
 * \param stops_blocks true to make this entity an obstacle for the blocks and statues
 */
Wall::Wall(
    const std::string& name,
    Layer layer,
    int x,
    int y,
    int width,
    int height,
    bool stops_hero,
    bool stops_enemies,
    bool stops_npcs,
    bool stops_blocks):
  MapEntity(name, 0, layer, x, y, width, height),
  enabled(true),
  waiting_enabled(false) {

  entity_types_stopped[ENTITY_HERO] = stops_hero;
  entity_types_stopped[ENTITY_ENEMY] = stops_enemies;
  entity_types_stopped[ENTITY_NPC] = stops_npcs;
  entity_types_stopped[ENTITY_BLOCK] = stops_blocks;
}

/**
 * \brief Destructor.
 */
Wall::~Wall() {

}

/**
 * \brief Returns the type of entity.
 * \return the type of entity
 */
EntityType Wall::get_type() const {
  return ENTITY_WALL;
}

/**
 * \brief Returns whether entities of this type can be drawn.
 * \return true if this type of entity can be drawn
 */
bool Wall::can_be_drawn() const {
  return false;
}

/**
 * \brief Returns whether this entity is an obstacle for another one
 * when it is enabled.
 * \param other another entity
 * \return true if this entity is an obstacle for the other one
 */
bool Wall::is_obstacle_for(const MapEntity& other) const {

  std::map<EntityType, bool>::const_iterator it =
      entity_types_stopped.find(other.get_type());
  return it != entity_types_stopped.end() && it->second;
}

}

File: tools/editor/src/main/java/org/solarus/editor/entities/Wall.java
/*
 * Copyright (C) 2006-2014 Christopho, Solarus - http://www.solarus-games.org
 *
 * Solarus Quest Editor is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Solarus Quest Editor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.solarus.editor.entities;

import java.awt.*;

import org.solarus.editor.*;

/**
 * A wall is an invisible entity that is an obstacle for some other types of entities.
 */
public class Wall extends MapEntity {

    /**
     * Description of the default image representing this kind of entity.
     */
    public static final EntityImageDescription[] generalImageDescriptions = {
        new EntityImageDescription("entity_wall.png", 0, 0, 32, 32)
    };

    /**
     * Resizable image of a wall.
     */
    private static Image resizableWallImage;

    /**
     * Constructor.
     * @param map the map
     */
    public Wall(Map map) throws MapException {
        super(map, 16, 16);
    }

    /**
     * Returns whether or not the entity is currently resizable.
     * @return true
     */
    public boolean isResizable() {
        return true;
    }

    /**
     * Declares all properties specific to the current entity type and sets
     * their initial values.
     */
    public void createProperties() throws MapException {
        createBooleanProperty("stops_hero", false, true);
        createBooleanProperty("stops_enemies", false, true);
        createBooleanProperty("stops_npcs", false, true);
        createBooleanProperty("stops_blocks", false, true);
    }

    /**
     * Checks the specific properties.
     * @throws MapException if a property is not valid
     */
    public void checkProperties() throws MapException {

        if (!getBooleanProperty("stops_hero") &&
                !getBooleanProperty("stops_enemies") &&
                !getBooleanProperty("stops_npcs") &&
                !getBooleanProperty("stops_blocks")) {
            throw new MapException("This entity should be obstacle for at least one type of entity to have an effect");
        }
    }

    /**
     * Draws the entity on the map editor.
     * @param g graphic context
     * @param zoom zoom of the image (for example, 1: unchanged, 2: zoom of 200%)
     * @param showTransparency true to make transparent pixels,
     * false to replace them by a background color
     */
    public void paint(Graphics g, double zoom, boolean showTransparency) {

        if (resizableWallImage == null) {
            resizableWallImage = Project.getEditorImageOrEmpty("entity_wall_resizable.png");
        }

        Rectangle positionInMap = getPositionInMap();
        int x1 = (int) (positionInMap.x * zoom);
        int y1 = (int) (positionInMap.y * zoom);
        int w = (int) (positionInMap.width * zoom);
        int h = (int) (positionInMap.height * zoom);
        int x2 = x1 + w;
        int y2 = y1 + h;
        int step = (int) (8 * zoom);

        for (int x = x1; x < x2; x += step) {
            for (int y = y1; y < y2 ; y += step) {
                g.drawImage(resizableWallImage, x, y, x + step, y + step, 0, 0, 8, 8, null);
            }
        }
        drawEntityOutline(g, zoom, new Color(240, 142, 142));
    }
}

